### PR TITLE
Boston housing example

### PR DESCRIPTION
**Pull request by MissingNO57**

### CODE REVIEW

File: libs/ml/examples/boston_housing_distributed_learning/main.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+              << std::endl;
+    return 1;
+  }
+
+  CoordinatorParams      coord_params;
+  ClientParams<DataType> client_params;
+
+  SizeType number_of_clients    = 10;
+  SizeType number_of_rounds     = 100;
+  coord_params.mode             = CoordinatorMode::SEMI_SYNCHRONOUS;
```
Comment (baykaner) [inline libs/ml/examples/boston_housing_distributed_learning/main.cpp]: before we put in this PR; can you try changing this to ASYNCHRONOUS, and running the example for 10 epochs with perhaps 4 clients (split the data into 4 groups)? And then lets take a look at plots of loss for all 4 clients.
Also add a test/evaluate method which shows performance on the full data set? and we can see what that looks like for all 4 clients as well.

File: libs/ml/examples/boston_housing_distributed_learning/main.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+#include "ml/distributed_learning/coordinator.hpp"
+#include "ml/distributed_learning/distributed_learning_client.hpp"
+#include "ml/ops/loss_functions/cross_entropy_loss.hpp"
+#include "ml/optimisation/adam_optimiser.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <ctime>
+#include <iostream>
+#include <ml/dataloaders/tensor_dataloader.hpp>
```
Comment (baykaner) [inline libs/ml/examples/boston_housing_distributed_learning/main.cpp]: "" not <>

File: libs/ml/examples/boston_housing_distributed_learning/main.cpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  {
+    if (i == number_of_clients - 1)
+    {
+      current_client_data_size = data_size - index;
+    }
+    splitting_points.push_back(current_client_data_size);
+    index += client_data_size;
+  }
+
+  std::vector<TensorType> data_tensors  = TensorType::Split(data_tensor, splitting_points, 1);
```
Comment (baykaner) [inline libs/ml/examples/boston_housing_distributed_learning/main.cpp]: separate function for splitting the data makes the example clearer

File: libs/ml/examples/boston_housing_distributed_learning/main.cpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    if (coordinator->GetMode() == CoordinatorMode::ASYNCHRONOUS)
+    {
+      continue;
+    }
+
+    // Synchronize weights by giving all clients average of all client's weights
+    VectorTensorType new_weights = clients[0]->GetWeights();
+
+    // Sum all weights
+    for (SizeType i{1}; i < number_of_clients; ++i)
```
Comment (baykaner) [inline libs/ml/examples/boston_housing_distributed_learning/main.cpp]: separate functions for summing all weights and dividing weights can also make the main loop clearer

File: libs/ml/include/ml/dataloaders/mnist_loaders/mnist_loader.hpp [hunk 5]
```diff
@@ -293,29 +293,42 @@ class MNISTLoader : public DataLoader<LabelType, InputType>
 
   void UpdateCursor() override
   {
-    if (this->mode_ == DataLoaderMode::TRAIN)
+    switch (this->mode_)
     {
+    case DataLoaderMode::TRAIN:
       this->current_cursor_ = train_cursor_;
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/mnist_loaders/mnist_loader.hpp:300]: curly braces for each case would be nice

**Review by baykaner — CHANGES_REQUESTED** [review]

**Review by baykaner — DISMISSED** [review]

File: libs/ml/examples/boston_housing_distributed_learning/main.cpp [hunk 6]
```diff
@@ -123,6 +123,50 @@ std::vector<TensorType> Split(TensorType &data, SizeType number_of_parts)
   return TensorType::Split(data, splitting_points, axis);
 }
 
+void Shuffle(TensorType &data, TensorType &labels)
+{
+  TensorType data_out   = data.Copy();
+  TensorType labels_out = labels.Copy();
+
+  std::vector<SizeType> indicies;
```
Comment (baykaner) [inline libs/ml/examples/boston_housing_distributed_learning/main.cpp]: typo: indices

**Review by baykaner — DISMISSED** [review]

**Review by baykaner — APPROVED** [review]
